indico/modules/events/timetable/controllers/display.py

```diff
@@ -73,28 +80,64 @@ def _process(self):
         return jsonify(html=html)
 
 
+@dataclass(frozen=True)
+class TimetableExportConfig:
+    show_title: bool
+    show_affiliation: bool
+    show_cover_page: bool
+    show_toc: bool
+    show_session_toc: bool
+    show_abstract: bool
+    dont_show_poster_abstract: bool
+    show_contribs: bool
+    show_length_contribs: bool
+    show_breaks: bool
+    new_page_per_session: bool
+    show_session_description: bool
+    print_date_close_to_sessions: bool
+
+
 class RHTimetableExportPDF(RHTimetableProtectionBase):
-    def _process(self):
+    @use_kwargs({'download': fields.Bool(load_default=False)}, location='query')
+    def _process(self, download):
         form = TimetablePDFExportForm(formdata=request.args, csrf_enabled=False)
+
         if form.validate_on_submit():
-            form_data = form.data_for_format
-            pdf_format = TimetablePDFFormat(form_data)
-            if not form.advanced.data:
-                pdf_format.contribsAtConfLevel = True
-                pdf_format.breaksAtConfLevel = True
-                pdf_class = SimplifiedTimeTablePlain
-                additional_params = {}
-            else:
-                pdf_class = TimeTablePlain
-                additional_params = {'firstPageNumber': form.firstPageNumber.data,
-                                     'showSpeakerAffiliation': form_data['showSpeakerAffiliation'],
-                                     'showSessionDescription': form_data['showSessionDescription']}
-            if request.args.get('download') == '1':
-                pdf = pdf_class(self.event, session.user, sortingCrit=None, ttPDFFormat=pdf_format,
-                                pagesize=form.pagesize.data, **additional_params)
-                return send_file('timetable.pdf', BytesIO(pdf.getPDFBin()), 'application/pdf')
+            days = {}
+            now = now_utc()
+            css = render_template('events/timetable/pdf/timetable.css')
+
+            for day in self.event.iter_days():
+                entries = (self.event.timetable_entries
```

Comment:
You're querying in a loop, and there are many more things getting queried later when displaying everything. For a large conference (CHEP 2018) this gives me 1.3k queries.

Have a look at the query in `inject_meeting_body`. You could probably extract this into a util function, and use it in both places. It does exactly what you need: Querying the full nested event timetable in a single query.

It may contain some more things than you need for the PDF, such as notes. You could avoid this by making it configurable when calling that function (e.g. `get_nested_timetable(event, *, include_notes=True, ...)`), but it's probably not a big deal since typically those large events do not have notes anyway.

---

In fact, the util could probably contain more than just the query. Even the logic in the loop could likely be reused. Just have some reasonable defaults, and it will work nicely for both places:

```py
def get_nested_timetable(event, *, include_notes=True, show_date='all', show_session='all', detail_level='all'):
    ...
```

By reusing all the code, you also get the `show_children_location` stuff we talked about on MM for free. I would probably take everything from querying w/ all the options up to including creating the `days` list and put that in the function, and then have it `return days, show_children_location, show_siblings_location` - or maybe make it a dataclass so it's a bit nicer than a tuple?

Reply:
Extracted and tested in the places that use the get_nested_timetable fn!